lib/fog/openstack.rb

```diff
@@ -129,7 +129,7 @@ def self.authenticate(options, connection_options = {})
       when /v2(\.\d+)?/
         authenticate_v2(options, connection_options)
       when /v3(\.\d+)?/
-        authenticate_v3(options, connection_options)
+        validate_or_authenticate_v3(options, connection_options)
```

Comment:
I believe there should be 2 different methods: #validate and #authenticate_v3.
And add any logic at this level if needed.